modules/search-as-you-type/src/main/java/org/elasticsearch/index/mapper/SearchAsYouTypeFieldMapper.java

```diff
@@ -0,0 +1,432 @@
+                .map(suggesterizedFieldType -> new SuggesterizedFieldMapper(suggesterizedFieldType, context.indexSettings()))
+                .collect(Collectors.toSet());
+            return new SearchAsYouTypeFieldMapper(
+                name(),
+                fieldType(),
+                suggesterizedFieldMappers,
+                context.indexSettings(),
+                multiFieldsBuilder.build(this, context),
+                copyTo
+            );
+        }
+    }
+
+    public static class SearchAsYouTypeAnalyzer extends AnalyzerWrapper {
+
+        private final Analyzer delegate;
+        private final boolean withShingles;
+        private final int shingleSize;
+        private final boolean withEdgeNGrams;
+
+        private SearchAsYouTypeAnalyzer(Analyzer delegate,
+                                        boolean withShingles,
+                                        int shingleSize,
+                                        boolean withEdgeNGrams) {
+
+            super(delegate.getReuseStrategy());
+            this.delegate = delegate;
+            this.withShingles = withShingles;
+            this.shingleSize = shingleSize;
+            this.withEdgeNGrams = withEdgeNGrams;
+        }
+
+        public static SearchAsYouTypeAnalyzer withShingles(Analyzer delegate, int shingleSize) {
+            return new SearchAsYouTypeAnalyzer(delegate, true, shingleSize, false);
+        }
+
+        public static SearchAsYouTypeAnalyzer withEdgeNGrams(Analyzer delegate) {
+            return new SearchAsYouTypeAnalyzer(delegate, false, -1, true);
+        }
+
+        public static SearchAsYouTypeAnalyzer withShinglesAndEdgeNGrams(Analyzer delegate, int shingleSize) {
+            return new SearchAsYouTypeAnalyzer(delegate, true, shingleSize, true);
+        }
+
+        @Override
+        protected Analyzer getWrappedAnalyzer(String fieldName) {
+            return delegate;
+        }
+
+        @Override
+        protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComponents components) {
+            // TODO we must find a way to add the last unigram term (michael jackson -> jackson)
+            TokenStream tokenStream = components.getTokenStream();
+            if (withShingles) {
+                tokenStream = new FixedShingleFilter(tokenStream, shingleSize);
+            }
+            if (withEdgeNGrams) {
+                tokenStream = new EdgeNGramTokenFilter(tokenStream, Defaults.MIN_GRAM, Defaults.MAX_GRAM, true);
+            }
+            return new TokenStreamComponents(components.getTokenizer(), tokenStream);
+        }
+
+        public boolean isWithEdgeNGrams() {
+            return withEdgeNGrams;
+        }
+
+        public int getShingleSize() {
+            return shingleSize;
+        }
+
+        public boolean isWithShingles() {
+            return withShingles;
+        }
+    }
+
+    public static class SuggesterizedFieldType extends TermBasedFieldType {
```

Comment:
We need something to infer the type of terms that are indexed in this field (`max_shingle` and if it contains edge ngrams or not).